conf_file.cpp

```diff
@@ -16,268 +16,412 @@
-}
+    const char *end;
+    struct section *s;
 
-const char *ConfFile::next_section()
-{
-    if (!_current_section->next)
-        return NULL;
+    end = (char *)memchr(addr, ']', len);
+    if (!end) {
+        log_error("On file %s: Line %d: Unfinished section name. Expected ']'", filename, line);
+        return nullptr;
+    }
+
+    if ((size_t)(end - addr) < len - 1) {
+        log_error("On file %s: Line %d: Unexpected characters after session name", filename, line);
+        return nullptr;
+    }
+
+    s = _find_section(addr + 1, len - 2);
+    if (s)
+        return s;
 
-    _current_section = _current_section->next;
-    return _current_section->name;
+    log_debug("ConfFile: Adding section '%.*s'", (int)(len - 2), addr + 1);
+    _sections = new section{addr + 1, len - 2, nullptr, _sections};
+    return _sections;
 }
 
-int ConfFile::_add_section(char *entry, int line)
+int ConfFile::_add_config(struct section *s, const char *entry, size_t entry_len,
+                          const char *filename, int line)
 {
-    struct section *section;
+    const char *equal_pos;
+    struct config *c;
+    const char *key, *value;
+    size_t key_len, value_len;
 
-    char *end = strchr(entry, ']');
-    if (!end) {
-        log_error("On file %s: Line %d: Unfinished section name. Expected ']'", _filename, line);
+    if (!(equal_pos = (char *)memchr(entry, '=', entry_len))) {
+        log_error("On file %s: Line %d: Missing '=' on config", filename, line);
         return -EINVAL;
     }
 
-    if (*(end + 1) != '\0') {
-        log_error("On file %s: Line %d: Unexpected characters after session name", _filename, line);
+    if (equal_pos == entry) {
+        log_error("On file %s: Line %d: Missing name on config", filename, line);
         return -EINVAL;
     }
 
-    // So section name is the string between '[]'
-    *entry = ' ';
-    *end = '\0';
-    _trim(entry);
-
-    // Ensure section is not duplicated. If it is,
-    // just make parser add on known section
-    section = _sections;
-    while (section) {
-        if (strcaseeq(section->name, entry)) {
-            _current_section = section;
-            free(entry);
-            return 0;
-        }
-        section = section->next;
+    if (equal_pos == (entry + entry_len - 1)) {
+        log_error("On file %s: Line %d: Missing value on config", filename, line);
+        return -EINVAL;
     }
 
-    // Section is new.
-    section = (struct section *)calloc(1, sizeof(struct section));
-    assert_or_return(section, -ENOMEM);
+    key = entry;
+    key_len = equal_pos - entry;
+    value = equal_pos + 1;
+    value_len = entry_len - key_len - 1;
 
-    section->name = entry;
-    section->next = _sections;
-    _sections = section;
+    _trim(&key, &key_len);
+    _trim(&value, &value_len);
 
-    _current_section = section;
+    c = _find_config(s, key, key_len);
+    if (c) {
+        c->value = value;
+        c->value_len = value_len;
+    } else
```

Comment:
Unbalanced braces should be avoided on mavlink-router - couldn't find a way to check for them on checkpatch..